### PR TITLE
Added "Carthage" package manager support

### PR DESCRIPTION
**Pull request by dodikk**
* Added iOS dynamic framework target
* Added iOS dynamic framework shared scheme
* Added a protocol for mocking the web socket in user's unit tests
* Introduced more strict warning level

### CODE REVIEW

**Comment by shaneneuerburg** [comment]
:+1:

**Comment by pettigree** [comment]
Is this something that is happening? It would be pretty cool to get this merged in.

**Comment by dfed** [comment]
We would love to get Carthage support in. Things that need to be done first:
1) @dodikk needs to sign our [CLA](https://docs.google.com/a/squareup.com/forms/d/13WR8m5uZ2nAkJH41k7GdVBXAAbzDk00vxtEYjd6Imzg/viewform?formkey=dDViT2xzUHAwRkI3X3k5Z0lQM091OGc6MQ&ndplr=1)
2) rebase master to resolve conflicts
3) Address review comments (incoming)

File: SocketRocket-iOSTests/SocketRocket_iOSTests.m [hunk 1]
```diff
@@ -0,0 +1,40 @@
+//
```
Comment (dfed) [inline SocketRocket-iOSTests/SocketRocket_iOSTests.m]: Please remove this file. It seems like it is boilerplate.

File: SocketRocket.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -621,11 +898,18 @@
 				ALWAYS_SEARCH_USER_PATHS = NO;
 				CLANG_CXX_LANGUAGE_STANDARD = "c++0x";
 				CLANG_ENABLE_OBJC_ARC = YES;
+				CLANG_WARN_ASSIGN_ENUM = YES;
```
Comment (dfed) [inline SocketRocket.xcodeproj/project.pbxproj]: why are these WARN changes necessary? Do they introduce new warnings?

File: SocketRocket.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -640,13 +924,14 @@
...
+				GCC_WARN_ABOUT_RETURN_TYPE = YES_ERROR;
+				GCC_WARN_FOUR_CHARACTER_CONSTANTS = YES;
 				GCC_WARN_SIGN_COMPARE = YES;
 				GCC_WARN_UNDECLARED_SELECTOR = YES;
-				GCC_WARN_UNINITIALIZED_AUTOS = YES;
+				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
-				IPHONEOS_DEPLOYMENT_TARGET = 5.1;
+				IPHONEOS_DEPLOYMENT_TARGET = 7.0;
```
Comment (dfed) [inline SocketRocket.xcodeproj/project.pbxproj]: This change looks like it changes the deployment target of the existing static lib. We should be creating a new target for Carthage so we can continue supporting old operating systems with the static lib.

File: SocketRocket.xcodeproj/project.xcworkspace/xcshareddata/SocketRocket.xccheckout [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dfed) [inline SocketRocket.xcodeproj/project.xcworkspace/xcshareddata/SocketRocket.xccheckout]: Please remove this file from your commit.

File: SocketRocket/SRWebSocket.h [hunk 5]
```diff
@@ -14,6 +14,7 @@
 //   limitations under the License.
 //
 
+#import <SocketRocket/SRWebSocketProtocol.h>
```
Comment (dfed) [inline SocketRocket/SRWebSocket.h]: nit: please order this import alphabetically

File: SocketRocket/SRWebSocket.h [hunk 6]
```diff
@@ -48,7 +49,7 @@ extern NSString *const SRHTTPResponseErrorKey;
 
 #pragma mark - SRWebSocket
 
-@interface SRWebSocket : NSObject <NSStreamDelegate>
+@interface SRWebSocket : NSObject <SRWebSocketProtocol, NSStreamDelegate>
```
Comment (dfed) [inline SocketRocket/SRWebSocket.h]: nit: please order protocols alphabetically

File: SocketRocket/SRWebSocket.m [hunk 7]
```diff
@@ -487,10 +490,11 @@ - (void)_readHTTPHeader;
     }
                         
     [self _readUntilHeaderCompleteWithCallback:^(SRWebSocket *self,  NSData *data) {
-        CFHTTPMessageAppendBytes(_receivedHTTPHeaders, (const UInt8 *)data.bytes, data.length);
+        CFIndex numBytes = (CFIndex)data.length;
+        CFHTTPMessageAppendBytes(self->_receivedHTTPHeaders, (const UInt8 *)data.bytes, numBytes);
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: why is this change necessary?

File: SocketRocket/SRWebSocket.m [hunk 8]
```diff
@@ -487,10 +490,11 @@ - (void)_readHTTPHeader;
...
         
-        if (CFHTTPMessageIsHeaderComplete(_receivedHTTPHeaders)) {
-            SRFastLog(@"Finished reading headers %@", CFBridgingRelease(CFHTTPMessageCopyAllHeaderFields(_receivedHTTPHeaders)));
+        if (CFHTTPMessageIsHeaderComplete(self->_receivedHTTPHeaders)) {
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: same q here.

File: SocketRocket/SRWebSocket.m [hunk 9]
```diff
@@ -487,10 +490,11 @@ - (void)_readHTTPHeader;
...
+            SRFastLog(@"Finished reading headers %@", CFBridgingRelease(CFHTTPMessageCopyAllHeaderFields(self->_receivedHTTPHeaders)));
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: and here

File: SocketRocket/SRWebSocket.m [hunk 10]
```diff
@@ -508,9 +512,12 @@ - (void)didConnect
         
     NSMutableData *keyBytes = [[NSMutableData alloc] initWithLength:16];
     SecRandomCopyBytes(kSecRandomDefault, keyBytes.length, keyBytes.mutableBytes);
-    
+
     if ([keyBytes respondsToSelector:@selector(base64EncodedStringWithOptions:)]) {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wassign-enum"
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: :+1:

File: SocketRocket/SRWebSocket.m [hunk 11]
```diff
@@ -671,7 +678,7 @@ - (void)_closeWithProtocolError:(NSString *)message;
     // Need to shunt this on the _callbackQueue first to see if they received any messages 
     [self _performDelegateBlock:^{
         [self closeWithCode:SRStatusCodeProtocolError reason:message];
-        dispatch_async(_workQueue, ^{
+        dispatch_async(self->_workQueue, ^{
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: again, the `self->` changes do not seem like they are necessary. Please remove them from your PR.

File: SocketRocket/SRWebSocket.m [hunk 12]
```diff
@@ -1087,7 +1094,7 @@ - (void)_pumpWriting;
              return;
         }
         
-        _outputBufferOffset += bytesWritten;
+        _outputBufferOffset += @(bytesWritten).unsignedIntegerValue;
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: Why is this change necessary?

File: SocketRocket/SRWebSocket.m [hunk 13]
```diff
@@ -1261,7 +1268,7 @@ - (BOOL)_innerPumpScanner {
                         });
                         return didWork;
                     } else {
-                        _currentStringScanPosition += valid_utf8_size;
+                        _currentStringScanPosition += (uint32_t)valid_utf8_size;
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: Why do we need this cast?

File: SocketRocket/SRWebSocket.m [hunk 14]
```diff
@@ -1677,7 +1685,8 @@ static inline int32_t validate_dispatch_data_partial_string(NSData *data) {
         }
     }
     
-    if (size != -1 && ![[NSString alloc] initWithBytesNoCopy:(char *)[data bytes] length:size encoding:NSUTF8StringEncoding freeWhenDone:NO]) {
```
Comment (dfed) [inline SocketRocket/SRWebSocket.m]: was the lack of cast causing a compiler warning?

**Comment by dfed** [comment]
Just did a review pass. I'd much prefer if you could have separate PRs for separate issues. Say, a PR enabling more warnings checks and resolving the warnings followed by a PR adding Carthage support.

File: SocketRocket/SRWebSocketProtocol.h [hunk 15]
```diff
@@ -0,0 +1,21 @@
+//
+//  SRWebSocketProtocol.h
```
Comment (dfed) [inline SocketRocket/SRWebSocketProtocol.h]: Can you explain why the creation of `SRWebSocketProtocol.h` was necessary?

**Comment by dfed** [comment]
Closing this out since we haven't heard back, and this PR addresses more issues in a single PR than we're comfortable merging. That said, the fixes here are good and we'd love to see this re-opened as multiple PRs with the above comments addressed.
